packages/liquidator/src/liquidationStrategy.js

```diff
@@ -0,0 +1,254 @@
+    assert(currentBlockTime >= 0, "requires currentBlockTime");
+    assert(position, "requires position");
+
+    // whale fund enabled
+    if (whaleDefenseFundWei && toBN(whaleDefenseFundWei).gt(toBN("0"))) {
+      // we should only liquidate the minimum here to extend withdraw
+      if (
+        shouldLiquidateMinimum({
+          position,
+          syntheticTokenBalance,
+          currentBlockTime
+        })
+      ) {
+        // we are using the WDF to liquidate. We need to alert user
+        emit("info", {
+          message:
+            "Liquidator bot is extending withdraw deadline, funds may need to be added to liquidate full position",
+          position: position,
+          maxLiquidationPrice: maxCollateralPerToken.toString(),
+          syntheticTokenBalance: syntheticTokenBalance.toString(),
+          maxTokensToLiquidateWei: maxTokensToLiquidateWei ? maxTokensToLiquidateWei.toString() : null,
+          whaleDefenseFundWei: whaleDefenseFundWei.toString(),
+          currentBlockTime,
+          ...logInfo
+        });
+        // this assume empMinSponsorSize will always be <= maxTokensToLiquidate
+        return createLiquidationParams({
+          sponsor: position.sponsor,
+          // Only liquidating the minimum size to extend withdraw
+          tokensToLiquidate: minSponsorSize,
+          currentBlockTime,
+          maxCollateralPerToken
+        });
+      }
+    }
+
+    // calculate tokens to liquidate respecting all constraints
+    const tokensToLiquidate = calculateTokensToLiquidate({
+      syntheticTokenBalance,
+      positionTokens: position.numTokens,
+      maxTokensToLiquidateWei
+    });
+
+    // check if we should try to liquidate this position
+    if (shouldLiquidate({ tokensToLiquidate })) {
+      // We should liquidate the entire position using our capital - wdf
+      return createLiquidationParams({
+        sponsor: position.sponsor,
+        tokensToLiquidate,
+        currentBlockTime,
+        maxCollateralPerToken
+      });
+    }
+    // we dont liquidate, dont return anything. Probably because amount to liquidate is 0
+  }
+
+  // Pure function which just calculates how far a timer has progressed between 0 and 100
+  // duration - total duration of the liveness in seconds
+  // end - end time that withdraw passes in unix seconds ts
+  // now - current block time in unix seconds ts
+  function withdrawProgressPercent(duration, end, now) {
+    const elapsed = end - now;
+    if (elapsed <= 0) return 100;
+    if (elapsed >= duration) return 0;
+    return 100 * (1 - elapsed / duration);
+  }
+
+  // Should we try to delay this withdrawal. Update this logic to change conditions to run delay strategy.
+  function shouldLiquidateMinimum({ position, syntheticTokenBalance, currentBlockTime }) {
+    assert(defenseActivationPercent >= 0, "requires defenseActivationPercent");
+    assert(liquidationLiveness > 0, "requires liquidationLiveness");
```

Comment:
nit: should these be checked at initialization time rather than call time so that the bot fails early rather than waiting until a liquidation is triggered to blow up? 

Are there any disadvantages to failing early here?

Reply:
Yes it should, and it does check on initialization. I've added additional checks within the function here in case you tried to call this without properly initializing the params on construction.

Reply:
I see. So, in particular, it seems like this checks to be sure you don't call this function manually in the case that WDF is NOT activated? That seems to be the only way you'd be able to get through the constructor without setting these. Is my thinking correct here?

Reply:
I think you are correct. You could probably simplify the asserts to check for that. IMO there's a potential refactor here to make all these functions pure and not rely on the closure, which probably would be better.